### PR TITLE
Upgrade Microsoft.CodeAnalysis.FxCopAnalyzers to 2.9.8

### PR DESCRIPTION
**Pull request by blairconrad**

### CODE REVIEW

**Comment by blairconrad** [comment]
I created separate commits this time because FxCop recommendations are often contentious!

File: src/FakeItEasy/Configuration/BuildableCallRule.cs [hunk 1]
```diff
@@ -77,7 +77,7 @@ public void UseApplicator(Action<IInterceptedFakeObjectCall> newApplicator)
         {
             if (this.wasApplicatorSet)
             {
-                throw new InvalidOperationException("The behavior for this call has already been defined");
+                throw new InvalidOperationException(ExceptionMessages.CallBehaviorAlreadyDefined);
```
Comment (blairconrad) [inline src/FakeItEasy/Configuration/BuildableCallRule.cs]: General comment on this commit: technically FxCop just complained about string literals in the constructors. I hunted down other constructed arguments and moved them to `ExceptionMessages` as well.

File: src/FakeItEasy.Analyzer.Shared/ArgumentConstraintAnalyzerBase.cs [hunk 2]
```diff
@@ -12,7 +12,7 @@ namespace FakeItEasy.Analyzer
     using Microsoft.CodeAnalysis.VisualBasic.Syntax;
 #endif
 
-    public abstract class ArgumentConstraintAnalyzerBase : DiagnosticAnalyzer
+    internal abstract class ArgumentConstraintAnalyzerBase : DiagnosticAnalyzer
```
Comment (blairconrad) [inline src/FakeItEasy.Analyzer.Shared/ArgumentConstraintAnalyzerBase.cs]: Other analyzers were internal, so I figured why not?

File: src/FakeItEasy.ruleset [hunk 3]
```diff
@@ -5,6 +5,7 @@
     <Rule Id="CA1020" Action="None" />
     <Rule Id="CA1200" Action="None" />
     <Rule Id="CA1305" Action="None" />
+    <Rule Id="CA2237" Action="None" /> <!-- Add [Serializable] to [...] as this type implements ISerializable -->
```
Comment (blairconrad) [inline src/FakeItEasy.ruleset]: Opinionated suppression: I don't see much point in marking our exceptions as serializable. Open to change.

Comment (thomaslevesque) [reply src/FakeItEasy.ruleset]: I'm in complete agreement

File: src/FakeItEasy.Analyzer.Shared/ArgumentConstraintAnalyzerBase.cs (hunk 2 above)
Comment (blairconrad) [reply src/FakeItEasy.Analyzer.Shared/ArgumentConstraintAnalyzerBase.cs]: Hmm. I only have 2017 at work, and it seems that _none_ of the analyzers work in 5.5.0. When I roll back to 4.9.2, `NonVirtualSetupAnalyzer`, which is internal works, as does `ArgumentConstraintOutsideCallSpecAnalyzer`, which is public.

Comment (thomaslevesque) [reply src/FakeItEasy.Analyzer.Shared/ArgumentConstraintAnalyzerBase.cs]: That's annoying... So, it doesn't matter that they're internal, but we must have broken something at some point. I'll try with VS2019.

Comment (thomaslevesque) [reply src/FakeItEasy.Analyzer.Shared/ArgumentConstraintAnalyzerBase.cs]: Works in VS2019.
We upgraded the analyzers to .NET Standard 2.0 in 5.4.0 (#1667); that's probably the cause. I didn't expect that they would stop working in VS2017... Well, people stuck on VS2017 can keep using an older version, I guess. But that means we probably should decouple the analyzer version from the library version...

File: src/FakeItEasy/ExceptionMessages.cs [hunk 4]
```diff
@@ -32,6 +33,27 @@ internal static class ExceptionMessages
...
         public static string NotAWrappingFake => "The configured fake is not a wrapping fake.";
 
+        public static string CallBehaviorAlreadyDefined =>
+            "The behavior for this call has already been defined";
+
+        public static string OutAndRefBehaviorAlreadyDefined =>
+            "How to assign out and ref parameters has already been defined for this call";
+
+        public static string NumberOfTimesNotGreaterThanZero =>
+            "The number of times to occur is not greater than zero.";
```
[the remark this answers is not in the document]
Comment (blairconrad) [reply src/FakeItEasy/ExceptionMessages.cs]: I wouldn't say "should", but "could" be, and I'll change if you really want, but I just transferred the existing message, and many of the others are phrased as complaints, like "The configured fake is not a wrapping fake." on line 34, which could be "The configured fake must be a wrapping fake."

[the remark this answers is not in the document]
Comment (thomaslevesque) [reply src/FakeItEasy/ExceptionMessages.cs]: You're right, of course. If we want to revisit the wording, we'll do it in another PR.

**Comment by blairconrad** [comment]
Thanks, @thomaslevesque. Squashed.

**Comment by thomaslevesque** [comment]
Thanks!

**Pull request merged**